File: python/full_demo.py
"""
Full SDK Demo

Demonstrates all SDK capabilities in one file.
"""

from hyperliquid_sdk import HyperliquidSDK, Order

# ============================================================================
# INITIALIZATION
# ============================================================================

# Simple init (uses PRIVATE_KEY env var)
sdk = HyperliquidSDK()

# Alternative: with auto-approval
# sdk = HyperliquidSDK(auto_approve=True)

# Alternative: explicit key
# sdk = HyperliquidSDK(private_key="0x...")

print(f"Wallet: {sdk.address}\n")

# ============================================================================
# MARKET DATA
# ============================================================================

print("=== Market Data ===")

mid = sdk.get_mid("BTC")
print(f"BTC mid: ${mid:,.2f}")

markets = sdk.markets()
print(f"Perp markets: {len(markets.get('perps', []))}")
print(f"Spot markets: {len(markets.get('spot', []))}")

dexes = sdk.dexes()
print(f"HIP-3 DEXes: {len(dexes.get('dexes', []))}")

# ============================================================================
# APPROVAL STATUS
# ============================================================================

print("\n=== Approval Status ===")

status = sdk.approval_status()
print(f"Approved: {status.get('approved', False)}")
if status.get("approved"):
    print(f"Max fee: {status.get('maxFeeRate')}")

# ============================================================================
# ORDER VALIDATION (PREFLIGHT)
# ============================================================================

print("\n=== Preflight Validation ===")

result = sdk.preflight("BTC", "buy", price=int(mid * 0.97), size=0.001)
print(f"Valid order: {result.get('valid', True)}")

# ============================================================================
# OPEN ORDERS
# ============================================================================

print("\n=== Open Orders ===")

orders = sdk.open_orders()
print(f"Open orders: {orders['count']}")

for o in orders["orders"][:3]:
    side = "BUY" if o["side"] == "B" else "SELL"
    print(f"  {o['name']} {side} {o['sz']} @ {o['limitPx']} (OID: {o['oid']})")

# ============================================================================
# ORDER PLACEMENT (commented to avoid real trades)
# ============================================================================

print("\n=== Order Methods ===")
print("Available order methods:")
print("  sdk.buy(asset, size=, notional=, price=, tif=)")
print("  sdk.sell(asset, size=, notional=, price=, tif=)")
print("  sdk.market_buy(asset, size=, notional=)")
print("  sdk.market_sell(asset, size=, notional=)")
print("  sdk.long(...)  # alias for buy")
print("  sdk.short(...) # alias for sell")
print("  sdk.order(Order.buy(...).size(...).price(...).gtc())")

# ============================================================================
# ORDER MANAGEMENT (commented to avoid real actions)
# ============================================================================

print("\n=== Order Management ===")
print("Available management methods:")
print("  order.cancel()                    # Cancel via order object")
print("  order.modify(price=, size=)       # Modify via order object")
print("  sdk.cancel(oid, asset=)           # Cancel by OID")
print("  sdk.cancel_by_cloid(cloid, asset) # Cancel by client order ID")
print("  sdk.cancel_all()                  # Cancel all orders")
print("  sdk.cancel_all('BTC')             # Cancel all BTC orders")
print("  sdk.schedule_cancel(time_ms)      # Dead-man's switch")
print("  sdk.close_position('BTC')         # Close a position")

# ============================================================================
# FLUENT ORDER BUILDER
# ============================================================================

print("\n=== Fluent Order Builder ===")
print("Order.buy('BTC')")
print("  .size(0.001)       # Size in asset units")
print("  .notional(100)     # Size in USD")
print("  .price(65000)      # Limit price")
print("  .gtc()             # Good Till Cancelled")
print("  .ioc()             # Immediate Or Cancel")
print("  .alo()             # Add Liquidity Only")
print("  .market()          # Market order")
print("  .reduce_only()     # Only close position")

print("\n=== Demo Complete ===")

File: rust/full_demo.rs
//! Full SDK Demo
//!
//! Demonstrates all SDK capabilities in one file.
//!
//! # Usage
//! ```bash
//! export PRIVATE_KEY=0x...
//! export ENDPOINT=https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN
//! cargo run --example full_demo
//! ```

use hyperliquid_sdk::HyperliquidSDK;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt::init();

    // ============================================================================
    // INITIALIZATION
    // ============================================================================

    let endpoint = std::env::var("ENDPOINT").ok();

    let mut builder = HyperliquidSDK::new();
    if let Some(ep) = endpoint {
        builder = builder.endpoint(ep);
    }

    // Alternative: with auto-approval
    // builder = builder.auto_approve(true);

    let sdk = builder.build().await?;

    println!("Wallet: {:?}\n", sdk.address());

    // ============================================================================
    // MARKET DATA
    // ============================================================================

    println!("=== Market Data ===");

    let mid = sdk.get_mid("BTC").await?;
    println!("BTC mid: ${:.2}", mid);

    let markets = sdk.markets().await?;
    if let Some(perps) = markets.get("perps").and_then(|v| v.as_array()) {
        println!("Perp markets: {}", perps.len());
    }
    if let Some(spot) = markets.get("spot").and_then(|v| v.as_array()) {
        println!("Spot markets: {}", spot.len());
    }

    let dexes = sdk.dexes().await?;
    if let Some(arr) = dexes.as_array() {
        println!("HIP-3 DEXes: {}", arr.len());
    }

    // ============================================================================
    // APPROVAL STATUS
    // ============================================================================

    println!("\n=== Approval Status ===");

    let status = sdk.approval_status().await?;
    let approved = status.get("approved").and_then(|v| v.as_bool()).unwrap_or(false);
    println!("Approved: {}", approved);
    if approved {
        if let Some(max_fee) = status.get("maxFeeRate").and_then(|v| v.as_str()) {
            println!("Max fee: {}", max_fee);
        }
    }

    // ============================================================================
    // ORDER VALIDATION (PREFLIGHT)
    // ============================================================================

    println!("\n=== Preflight Validation ===");

    let limit_price = (mid * 0.97) as i64;
    let result = sdk.preflight("BTC", "buy", Some(limit_price as f64), Some(0.001)).await?;
    let valid = result.get("valid").and_then(|v| v.as_bool()).unwrap_or(true);
    println!("Valid order: {}", valid);

    // ============================================================================
    // OPEN ORDERS
    // ============================================================================

    println!("\n=== Open Orders ===");

    let orders = sdk.open_orders().await?;
    if let Some(arr) = orders.as_array() {
        println!("Open orders: {}", arr.len());
        for o in arr.iter().take(3) {
            let name = o.get("coin").and_then(|v| v.as_str()).unwrap_or("?");
            let side = if o.get("side").and_then(|v| v.as_str()) == Some("B") { "BUY" } else { "SELL" };
            let sz = o.get("sz").and_then(|v| v.as_str()).unwrap_or("?");
            let px = o.get("limitPx").and_then(|v| v.as_str()).unwrap_or("?");
            let oid = o.get("oid").and_then(|v| v.as_u64()).unwrap_or(0);
            println!("  {} {} {} @ {} (OID: {})", name, side, sz, px, oid);
        }
    }

    // ============================================================================
    // ORDER PLACEMENT (commented to avoid real trades)
    // ============================================================================

    println!("\n=== Order Methods ===");
    println!("Available order methods:");
    println!("  sdk.market_buy(asset).await.notional(usd).await?");
    println!("  sdk.market_buy(asset).await.size(qty).await?");
    println!("  sdk.market_sell(asset).await.notional(usd).await?");
    println!("  sdk.limit_buy(asset, price, size).await?");
    println!("  sdk.limit_sell(asset, price, size).await?");
    println!("  sdk.order(Order::buy(...).size(...).price(...).gtc())");

    // ============================================================================
    // ORDER MANAGEMENT (commented to avoid real actions)
    // ============================================================================

    println!("\n=== Order Management ===");
    println!("Available management methods:");
    println!("  order.cancel().await?               // Cancel via order object");
    println!("  order.modify(price, size).await?    // Modify via order object");
    println!("  sdk.cancel(oid, asset).await?       // Cancel by OID");
    println!("  sdk.cancel_by_cloid(cloid, asset)?  // Cancel by client order ID");
    println!("  sdk.cancel_all(None).await?         // Cancel all orders");
    println!("  sdk.cancel_all(Some(\"BTC\")).await?  // Cancel all BTC orders");
    println!("  sdk.close_position(\"BTC\").await?    // Close a position");

    // ============================================================================
    // FLUENT ORDER BUILDER
    // ============================================================================

    println!("\n=== Fluent Order Builder ===");
    println!("Order::buy(\"BTC\")");
    println!("  .size(0.001)       // Size in asset units");
    println!("  .notional(100)     // Size in USD");
    println!("  .price(65000)      // Limit price");
    println!("  .gtc()             // Good Till Cancelled");
    println!("  .ioc()             // Immediate Or Cancel");
    println!("  .alo()             // Add Liquidity Only");
    println!("  .market()          // Market order");
    println!("  .reduce_only()     // Only close position");

    println!("\n=== Demo Complete ===");

    Ok(())
}
